docs/USAGE-GUIDE.md

```diff
@@ -607,66 +607,29 @@ In TLS1.3, session ticket messages are sent after the handshake as "post-handsha
 
-The callback can be be invoked in two modes
-- **S2N_CLIENT_HELLO_CB_BLOCKING** (default):
+`s2n_client_hello_get_cipher_suites()` will retrieve the list of cipher suites sent by the client.
 
-    In this mode s2n-tls expects the callback to complete its work
-    and return the appropriate response code before the handshake continues.
-    If any of the connection properties were changed based on the server_name
-    extension the callback must either return a value greater than 0 or invoke **s2n_connection_server_name_extension_used**,
-    otherwise the callback returns 0 to continue the handshake.
+`s2n_client_hello_get_session_id()` will get the session ID sent by the client in the ClientHello message. Note that this value may not be the session ID eventually associated with this particular connection since the session ID can change when the server sends the Server Hello. The official session ID can be retrieved with `s2n_connection_get_session_id()`.
 
-- **S2N_CLIENT_HELLO_CB_NONBLOCKING**:
+Call `s2n_client_hello_get_extensions()` to retrieve the entire list of extensions sent in the Client Hello. Calling `s2n_client_hello_get_extension_by_id()` allows you to interrogate the `s2n_client_hello` struct for a specific extension.
 
-    In non-blocking mode, s2n-tls expects the callback to not complete its work. If the callback
-    returns a response code of 0 s2n-tls will return **S2N_FAILURE** with **S2N_ERR_T_BLOCKED**
-    error type and **s2n_blocked_status** set to **S2N_BLOCKED_ON_APPLICATION_INPUT**.
-    The handshake is paused and further calls to **s2n_negotiate** will continue to return the
-    same error until **s2n_client_hello_cb_done** is invoked for the **s2n_connection** to resume
-    the handshake. This allows s2n-tls clients to process client_hello without
-    blocking and then resume the handshake at a later time.
-    If any of the connection properties were changed on the basis of the server_name extension then
-    **s2n_connection_server_name_extension_used** must be invoked before marking the callback done.
+### Client Hello Callback
 
-The callback can return a negative value to make s2n-tls terminate the
-handshake early with a fatal handshake failure alert.
+Users can gain access to the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()` with an implemented `s2n_client_hello_fn`. A common use-case for this is to switch the associated `s2n_config` based on information in the Client Hello. Note that `s2n_connection_server_name_extension_used()` MUST be invoked before exiting the callback if any of the connection properties were changed on the basis of the Server Name extension. If desired, the callback can return a negative value to make s2n-tls terminate the handshake early with a fatal handshake failure alert.
```

Comment:
I wouldn't say "common usecase", maybe "A possible usecase" :) We also might want a warning here that modifying the connection or config mid-handshake based on the ClientHello is potentially dangerous and should be done carefully. In particular, switching from a config that supports TLS1.3 to one that doesn't opens the server up to a possible version downgrade attack. See https://github.com/aws/s2n-tls/issues/2223#issuecomment-686231386.

Maybe we also want to open an issue to record "could have done TLS1.3" before the client hello callback and base our downgrade protection off of that instead of off the current config after the callback.